### PR TITLE
Text color map (Issue #173)

### PR DESCRIPTION
**Pull request by the-drunk-coder**
Please read the "Making a PR" section of [`CONTRIBUTING.md`](https://github.com/emilk/egui/blob/master/CONTRIBUTING.md) before opening a Pull Request!

### CODE REVIEW

File: epaint/src/shape.rs [hunk 1]
```diff
@@ -49,6 +49,15 @@ pub enum Shape {
         /// If true, tilt the letters for an ugly italics effect
         fake_italics: bool,
     },
+    MulticolorText {
+        /// Top left corner of the first character.
+        pos: Pos2,
+        /// The layed out text
+        galley: Galley,
+	text_style: TextStyle, // TODO: Font?
+	color_map: TextColorMap,
```
Comment (emilk) [inline epaint/src/shape.rs:58]: No need to add a separate `MulticolorText` - just add `color_map: TextColorMap` to `Text`, and set it to `TextColorMap::default()` when there are no color changes

File: epaint/src/tessellator.rs [hunk 2]
```diff
@@ -705,6 +725,74 @@ impl Tessellator {
         }
         assert_eq!(chars.next(), None);
     }
+    
+    pub fn tessellate_multicolor_text(
```
Comment (emilk) [inline epaint/src/tessellator.rs:729]: I rather have just one `tesselate_text`. With `TextColorMap::default()` they should act the same anyway.

**Review by emilk — CHANGES_REQUESTED** [review]
Thank you for this PR! Hot tip: configure your editor to run `cargo fmt` on save :)

**Comment by emilk** [comment]
I'll convert this to a draft and you can change it to "Ready for review" when it is green

**Comment by the-drunk-coder** [comment]
I removed the additional `tesselate_multicolor_text` etc. I only kept the multicolor galley because there were already some methods in place, so I thought that was in line ...

**Review by emilk — CHANGES_REQUESTED** [review]
This is starting to look good!

What I think is missing is some way to use the color map, and a demonstration of it in the demo app. I think at least `Label`, and maybe `TextEdit`, should have an option to set a `TextColorMap`. It would be great to demonstrate it with some simple demo, at least for `Label`. It can be added to `egui_demo_lib/src/apps/demo/demo_window.rs`.

...or if you are feeling very ambitious, check out the syntax highlighting in `egui_demo_lib/src/apps/http_app.rs` which currently uses a bunch of labels, but could be just a single `Label` with a `TextColorMap`. To test that use `cargo run --all-features -p egui_demo_app` (you need to rebase on `master` first). Totally optional though!

**Comment by emilk** [comment]
Oh, and please add a line to `CHANGELOG.md`!

Sorry for the slow review btw

**Comment by the-drunk-coder** [comment]
No worries, got plenty of things on my plate as well. 

So, would you prefer having separate methods for the galleys in `Painter` ? Like, `galley_with_italics` and `galley_with_italics_multicolor` ? Or rather just one ? There's plenty of galley methods, so I'm not quite sure here ... personally I'd make separate ones ...

**Comment by the-drunk-coder** [comment]
Also, now that I think about it, at least for `TextEdit`, I'm not sure whether it makes sense to set a static color map ? I mean, the way I use it it's generated dynamically based on the content.

File: epaint/src/text/fonts.rs [hunk 3]
```diff
@@ -139,7 +142,7 @@ impl Default for FontDefinitions {
             fonts_for_family.insert(
                 FontFamily::Monospace,
                 vec![
-                    "ProggyClean".to_owned(),
+                    "ComicMono".to_owned(),
```
Comment (emilk) [inline epaint/src/text/fonts.rs:145]: please revert!

File: egui/src/painter.rs [hunk 4]
```diff
@@ -3,9 +3,9 @@ use crate::{
     layers::{LayerId, PaintList, ShapeIdx},
     Color32, CtxRef,
 };
-use epaint::{
+use epaint::{
```
Comment (emilk) [inline egui/src/painter.rs:6]: Configure your editor to trim trailing spaces!

**Review by emilk — COMMENTED** [review]
This needs a demo!

For instance, if `Label` was given a method to set a color map, then `egui_demo_lib/src/apps/http_app.rs` could be configured to use it for syntax highlighting (requires `syntect` feature) in `ColoredText::ui`.

**Comment by emilk** [comment]
Closed in favor of https://github.com/emilk/egui/pull/682 which is a complete rewrite of the text layout engine which supports color ranges and much, much more.

**Pull request closed without merging**